yaqd_andor/_andorsdk2_ixon.py

```diff
@@ -0,0 +1,309 @@
+        )
+        self.logger.info(
+            f"binning={binning}, left_aoi=pixel {left}, right_aoi=pixel {right}, top_aoi=pixel {top}, bottom_aoi=pixel {bottom}."
+        )
+        if code != 20002:
+            raise ValueError(str(self.errorlookup(code)))
+
+        else:
+            self.buffer_size = int(int(width) * int(height))
+            self.buffer = np.zeros([arrwidth, arrheight], dtype=int)
+            self.buffer = np.ascontiguousarray(self.buffer, dtype=int)
+
+            # apply shape, mapping
+            self._channel_shapes = {"image": (int(height), int(width))}
+            self.x_ai = np.arange(int(left), int(left + width * binning), int(binning), dtype=int)[
+                None, :
+            ]
+            self.y_ai = np.arange(int(top), int(top + height * binning), int(binning), dtype=int)[
+                :, None
+            ]
+            self.binning = int(binning)
+            return
+
+    def _set_temperature(self):
+        self.sensor_cooling = self._config["sensor_cooling"]
+        if self.sensor_cooling:
+            self.sensor_temp_control = int(self._config["sensor_temperature"])
+            code = self.sdk.CoolerON()
+            if code != 20002:
+                raise ValueError(str(self.errorlookup(code)))
+            code = self.sdk.SetTemperature(self.sensor_temp_control)
+            if code != 20002:
+                raise ValueError(str(self.errorlookup(code)))
+            self.logger.info(f"Sensor is cooling.  Target temp is {self.sensor_temp_control} C.")
+            self._loop.run_in_executor(None, self._check_temp_stabilized)
+        else:
+            code, self.sensor_temp = self.sdk.GetTemperature()
+            self.logger.info(f"Sensor is not cooled.  Current temp is {self.sensor_temp} C.")
+
+    def _check_temp_stabilized(self):
+        code, self.sensor_temp = self.sdk.GetTemperature()
+        diff = float(self.sensor_temp_control) - float(self.sensor_temp)
+        while (
+            np.abs(diff) > 1.0 and self.stop_update is False
+        ):  # this is a tolerance, and is subject to change
+            self.logger.info(
+                f"Sensor is cooling.  Target: {self.sensor_temp_control} C.  Current: {self.sensor_temp:0.2f} C."
+            )
+            sleep(3)
+            code, self.sensor_temp = self.sdk.GetTemperature()
+            diff = float(self.sensor_temp) - float(self.sensor_temp_control)
+        if np.abs(diff) <= 1.0:
+            self.logger.info("Sensor temp is stabilized.")
+
+    def get_dependent_hardware(self):
+        dependents = dict()
+        if self.spec_client is not None:
+            dependents = {"monochromator": self._spec_position}
+        return dependents
+
+    def set_exposure_time(self, exposure_time):
+        # Sets the exposure time in seconds (float)
+        self.stop_update = True
+        while self._busy == True:
+            sleep(0.10)
```

Comment:
this will block code execution if busy is true

Reply:
I think the camera typically does not accept commands when it is processing something else such as a measurement or a status report.   I have seen some bad behavior by the camera when feeding commands to it when it is in the process of another command.   Unfortunately the permutations of commands is too large to go through to see when this happens.  Sounds like an event loop or some such involving asynchronous loops would be needed.   I was hoping that the limited features being used would circumvent need for async loops.

Reply:
the problem is this code is not doing what you think it is doing:
* when _busy == False, there is no sleep
* when _busy == True, this process is an infinite loop

Reply:
It might be illustrative to try using this code to test your functionality:
```
self._busy = True
while self._busy:
    sleep(0.1)
    self.logger.info("stuck in loop")
```

Reply:
ok it took a while to figure out what was wrong because I thought update_state was performing an operation when self.stop_update was set to True but it is not